Validate Painless Lab index field

cjcenizal opened the pull request:
Fixes https://github.com/elastic/kibana/issues/60020

This required significant refactoring of the store's shape. To make this refactor easier, I also tried to simplify the way types and constants are organized. An additional improvement I made was to clarify the labeling of various form fields.

![image](https://user-images.githubusercontent.com/1238659/77215676-ba548880-6ad2-11ea-9e3b-7c19fbfec798.png)

elasticmachine commented:
Pinging @elastic/es-ui (Team:Elasticsearch UI)

cjcenizal commented on x-pack/plugins/painless_lab/public/application/context.tsx:
```diff
@@ -8,50 +8,89 @@ import React, { createContext, ReactNode, useState, useContext } from 'react';
-  }));
+  };
 
-  const updateState = (getNextState: (s: Store) => Partial<Store>): void => {
```
@jloleysens Was there a particular reason to accept a callback here instead of the state change?

jloleysens replied:
The primary reason was to provide callers of `updateState` with a reference to the previous state so they can make decisions about it if needed, e.g., `updateState(s => s.something ? ...  : ...)` and so that it is a bit more convenient to work with as the store grows in complexity (otherwise we have to fetch the store from `const { store } = useAppContext()` and pass use it before we update something. So it is primarily a mechanism of optimisation.

This followed the `useState` pattern, but I am happy to have it accept both a callback and an object :+1:

cjcenizal replied:
OK, thanks for explaining! I follow your reasoning until this part:

> otherwise we have to fetch the store from const { store } = useAppContext() and use it when we update something and then update it again in the future if the shape changes.

This is where I get lost. Is there an example of this in the codebase? Philosophically I would suggest waiting until this becomes a problem before introducing a solution for it.

cjcenizal commented on x-pack/plugins/painless_lab/public/application/context.tsx:
```diff
@@ -8,50 +8,89 @@ import React, { createContext, ReactNode, useState, useContext } from 'react';
   links: Links;
 }
 
+const initialPayload = {
```
@jloleysens What do you think of bringing this into this file? It's not consumed elsewhere so I think it creates noise to extract it. If we want to break this file's code up into multiple files, I'd prefer to break it out into its own module (a directory with an index file) to make it clear what's intended for consumption elsewhere.

jloleysens replied:
Sure, I am happy with this change. I had the `store.ts` file separate so that if we choose to use a reducer in future we can just add the reducer to the store file.

jloleysens replied:
I do think if we start to expand more on validators it would be nice to have a separate file for these. Ideally I like to keep the `context` file as minimal as possible so that it only has the concern setting up the context. Where, at the moment, it houses store setup and validation logic - perhaps worth creating the folder structure you were speaking about. That way future engineers know where to put things?

cjcenizal replied:
OK, I'll create a folder now. In terms of validation, I would expect us to try using the form hook lib if we want to add more rules in the future.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/35138/)
* Commit: 5aadb7764d1abe72e5363ffb0a184a96a63080b1

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/35138/","number":"35138","commit":"5aadb7764d1abe72e5363ffb0a184a96a63080b1"}]}
PIPELINE-->

jloleysens commented on x-pack/plugins/painless_lab/public/application/context.tsx:
```diff
@@ -8,50 +8,89 @@ import React, { createContext, ReactNode, useState, useContext } from 'react';
+  context: painlessContextOptions[0].value,
+  code: exampleScript,
+  parameters: `{
+  "string-parameter": "yay",
```
Should we clean these up?

jloleysens replied:
Or are they intended as the parameter defaults? `"yay"` seems a bit too informal to me 😄

cjcenizal replied:
Haha yes I intended them to be defaults. I'll try to find something appropriately solemn. :)

jloleysens commented on x-pack/plugins/painless_lab/public/application/context.tsx:
```diff
@@ -8,50 +8,89 @@ import React, { createContext, ReactNode, useState, useContext } from 'react';
-    ...initialState,
+  const PAINLESS_LAB_KEY = 'painlessLabState';
+
+  const defaultPayload = {
```
This should be inside of a getter like so that we have this pattern: `useState(() => ...)`. In this implementation we are calling `JSON.parse` and `localStorage.getItem` for every state update. This post explains the lazy init idea: https://dmitripavlutin.com/react-usestate-hook-guide/#3-lazy-initialization-of-state

cjcenizal replied:
Thanks for sharing that link! I was completely overlooking that performance hit and the link opened my eyes. Will fix.

cjcenizal commented:
Thanks for the review @jloleysens! I've addressed all of your feedback except for your suggestion to allow `updatePayload` to accept both a payload object and a callback. I'd be happy to implement this in a separate PR if you feel strongly about it.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/35597/)
* Commit: 87a7f587d4f1cd9ded959e0816716ecfeb27e049

### History
* :green_heart: [Build #35138](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/35138/) succeeded 5aadb7764d1abe72e5363ffb0a184a96a63080b1

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/35138/","number":"35138","commit":"5aadb7764d1abe72e5363ffb0a184a96a63080b1"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/35597/","number":"35597","commit":"87a7f587d4f1cd9ded959e0816716ecfeb27e049"}]}
PIPELINE-->
